boxtree/tree_build.py

```diff
@@ -46,6 +56,240 @@ class MaxLevelsExceeded(RuntimeError):
+    :arg refine_flags: a boolean array of size `nboxes`.
+    :arg coarsen_flags: a boolean array of size `nboxes`.
+    :returns: a processed copy of the tree.
+    """
+    nchildren = 2**tob.dim
+
+    if refine_flags is None:
+        refine_flags = np.zeros(tob.nboxes, bool)
+    if refine_flags[~tob.get_leaf_flags()].any():
+        raise ValueError("attempting to split non-leaf")
+
+    if coarsen_flags is None:
+        coarsen_flags = np.zeros(tob.nboxes, bool)
+    if coarsen_flags[~tob.get_leaf_flags()].any():
+        raise ValueError("attempting to coarsen non-leaf")
+
+    # ------------------- refinement always respects the flags
+
+    refine_parents, = np.where(refine_flags)
+    n_new_boxes = len(refine_parents) * nchildren
+    nboxes_new = tob.nboxes + n_new_boxes
+
+    child_box_starts = (
+            tob.nboxes
+            + nchildren * np.arange(len(refine_parents)))
+
+    refine_parents_per_child = np.empty(
+            (nchildren, len(refine_parents)),
+            np.intp)
+    refine_parents_per_child[:] = refine_parents.reshape(-1)
+    refine_parents_per_child = refine_parents_per_child.reshape(-1)
+
+    box_parents = resized_array(tob.box_parent_ids, nboxes_new)
+    box_centers = resized_array(tob.box_centers, nboxes_new)
+    box_children = resized_array(tob.box_child_ids, nboxes_new)
+    box_levels = resized_array(tob.box_levels, nboxes_new)
+
+    # new boxes are appended at the end, so coarsen_flags are still meaningful
+    box_parents[tob.nboxes:] = refine_parents_per_child
+    box_levels[tob.nboxes:] = tob.box_levels[box_parents[tob.nboxes:]] + 1
+    box_children[:, refine_parents] = (
+        child_box_starts
+        + np.arange(nchildren).reshape(-1, 1))
+
+    for i in range(2**tob.dim):
+        children_i = box_children[i, refine_parents]
+        offsets = (
+                tob.root_extent
+                * vec_of_signs(tob.dim, i).reshape(-1, 1)
+                * (1/2**(1+box_levels[children_i])))
+        box_centers[:, children_i] = (
+                box_centers[:, refine_parents]
+                + offsets)
+
+    # ------------------- coarsen only if all peers are leaves after refinement
+
+    coarsen_flags = resized_array(coarsen_flags, nboxes_new)
+    coarsen_sources, = np.where(coarsen_flags)
+    if coarsen_sources:
+        # FIXME
+        raise NotImplementedError
+        coarsen_parents = tob.box_parent_ids[coarsen_sources]
+        coarsen_peers = tob.box_child_ids[:, coarsen_parents]
+        coarsen_peer_is_leaf = tob.get_leaf_flags()[coarsen_peers]
+        coarsen_exec_flags = np.all(coarsen_peer_is_leaf, axis=0)
+
+        coarsen_flags_ignored = (coarsen_exec_flags != coarsen_flags)
+        if np.any(coarsen_flags_ignored):
+            import warnings
+            warnings.warn(f"{np.sum(coarsen_flags_ignored)} out of "
+                          f"{np.sum(coarsen_flags)} coarsening flags ignored "
+                          "to prevent removing non-leaf boxes")
```

Comment:
Maybe that should be an error?

Reply:
In practice, I think it would be more convenient if the function simply proceed after ignoring those flags. Then the user would not need to worry about level restrictions when setting coarsening flags.

Reply:
No user I know wants functions to simply ignore inputs they make. They can mask out leaves on their own IMO.

Reply:
I made throwing an error the default behavior, but still left a flag to turn it off. The rationale for ignoring some flags is that a PDE solver may compute the flags solely based on error estimates without minding the existence of level restrictions.

Reply:
([440bc2f](https://github.com/inducer/boxtree/pull/35/commits/440bc2f161fd09271d7f167fe834c05032beddf1))